File: 42. Trapping Rain Water.cpp
// 42. Trapping Rain Water

// 备忘法 从暴力法优化过来
// 暴力法中，我们在每一个位置都要求它到左端和右端的最大值，包含大量的重复计算 (其实从这点能看出可以使用单调栈来解)
// 备忘法用来记忆计算结果，避免重复计算

class Solution {
public:
    int trap(vector<int>& height) {
        int res = 0, len = height.size();
        if(height.size() < 3) return 0;
        
        vector<int> left_max(len), right_max(len);
        
        left_max[0] = height[0];
        for(int i=1; i<len; i++) {
            left_max[i] = max(height[i], left_max[i-1]);
        }
        right_max[len-1] = height[len-1];
        for(int i=len-2; i>=0; i--) {
            right_max[i] = max(height[i], right_max[i+1]);
        }
        
        for(int i=1; i<len-1; i++) { // 第一个和最后一个是无法承接雨水的
            res += min(left_max[i], right_max[i]) - height[i];
        }
        
        return res;
    }
};

File: 42. Trapping Rain Water.java
// 42. Trapping Rain Water

// 单调栈 递减栈
// 分析题意 不难看出我们需要找到当前元素的左上界 和 右上界 因而使用单调栈
// 这里使用递减栈 注意 一定要想清楚使用递增栈还是递减栈

class Solution {
    public int trap(int[] height) {
        int res = 0, idx = 0;
        Stack<Integer> stack = new Stack<>();
        
        while(idx < height.length) {
            while(!stack.isEmpty() && height[idx] > height[stack.peek()]) {
                int top = stack.pop();
                if(stack.isEmpty()) {
                    break;
                } else{
                    int dis = idx-stack.peek()-1;
                    int h = Math.min(height[idx], height[stack.peek()]) - height[top];
                    res += dis * h;
                }
            }
            stack.push(idx++);
        }
        return res;
    }
}

// for循环和单调栈模板代码更接近
class Solution {
    public int trap(int[] height) {
        int res = 0;
        Stack<Integer> stack = new Stack<>();
        
        for(int idx = 0; idx < height.length; idx++) {
            while(!stack.isEmpty() && height[idx] > height[stack.peek()]) { // 模板代码 维护一个递减栈 注意不是大于等于
                int top = stack.pop(); // 模板代码 维护一个递减栈
                if(stack.isEmpty()) {
                    break;
                } else{
                    int dis = idx-stack.peek()-1; // 上界本身不能存贮雨水
                    int h = Math.min(height[idx], height[stack.peek()]) - height[top];
                    res += dis * h;
                }
            }
            stack.push(idx); // 模板代码 维护一个递减栈
        }
        return res;
    }
}